### PR TITLE
feat(warden,axelar): WardenHandler

### PR DESCRIPTION
**Pull request by rudewalt**
…g with Axelar and handling reqeusts from Axelar relayers

### CODE REVIEW

File: contracts/interfaces/IAxelarGateway.sol [hunk 1]
```diff
@@ -0,0 +1,35 @@
+// SPDX-License-Identifier: GPL-3.0
+pragma solidity ^0.8.0;
```
Comment (Overseven) [inline contracts/interfaces/IAxelarGateway.sol:2]: `pragma solidity 0.8.26;`

File: contracts/WardenHandler.sol [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+  error InvalidSourceChain();
+  error InvalidActionType();
+  error AmountTooBig();
+
+  uint8 private constant STAKE_ACTION_TYPE = 0;
+  uint8 private constant UUNSTAKE_ACTION_TYPE = 1;
+  uint8 private constant REINIT_ACTION_TYPE = 2;
+
+  uint8 internal constant SUCCESS_STATUS = 0;
+  uint8 internal constant ERROR_STATUS = 1;
```
Comment (optifat) [inline contracts/WardenHandler.sol:24]: enums?

File: contracts/WardenHandler.sol [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+  /// @param axelarGasService Address of Axelar gas service
+  /// @param wardenChain Identifier of warden chain
+  /// @param wardenContractAddress Contract address in warden chain
+  function __WardenHandler_init(
+    address axelarGateway,
+    address axelarGasService,
+    string calldata wardenChain,
+    string calldata wardenContractAddress
+  ) internal onlyInitializing {
+    if (axelarGateway == address(0)) revert InvalidAddress();
```
Comment (optifat) [inline contracts/WardenHandler.sol:74]: can `axelarGasService` be `address(0)`?

File: contracts/WardenHandler.sol [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+/// @notice Handles requests from Warden
+abstract contract WardenHandler is Initializable {
+  error NotApprovedByGateway();
+  error InvalidAddress();
+  error InvalidSourceChain();
+  error InvalidActionType();
+  error AmountTooBig();
+
+  uint8 private constant STAKE_ACTION_TYPE = 0;
+  uint8 private constant UUNSTAKE_ACTION_TYPE = 1;
```
Comment (optifat) [inline contracts/WardenHandler.sol:20]: two u's -- is it a typo?

File: contracts/WardenHandler.sol [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+    WardenHandlerData storage $ = _getWardenHandlerData();
+
+    string memory wardenChain = $.wardenChain;
+    if (!Strings.equal(wardenChain, sourceChain)) revert InvalidSourceChain();
+
+    string memory wardenContractAddress = $.wardenContractAddress;
+    if (!Strings.equal(wardenContractAddress, sourceAddress)) revert InvalidSourceChain();
+
+    IAxelarGateway gateway = IAxelarGateway($.axelarGateway);
+    if (!gateway.validateContractCall(commandId, sourceChain, sourceAddress, keccak256(payload)))
```
Comment (optifat) [inline contracts/WardenHandler.sol:186]: Wouldn't it be better to `gateway.validateContractCall(commandId, $.wardenChain, $.wardenContractAddress, keccak256(payload))`? Seems like you can skip all the previous checks if done like this

File: contracts/WardenHandler.sol [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+
+      tokenAddress = reinitResult.tokenAddress;
+      tokenAmount = reinitResult.tokenAmount;
+      response = _createReinitResponse(reinitResult.reinitUnstakeId);
+    } else {
+      revert InvalidActionType();
+    }
+
+    if (tokenAmount != 0) {
+      IERC20(tokenAddress).approve(address(gateway), tokenAmount);
```
Comment (optifat) [inline contracts/WardenHandler.sol:215]: `forceApprove` here and in all similar places

File: contracts/WardenHandler.sol [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+  /// @param unstakeId Unstake identifier
+  /// @param lpAmount Amount of lp token to unstake
+  /// @return ustake result
+  function _handleUnstakeRequest(uint64 unstakeId, uint128 lpAmount) internal virtual returns (UnstakeResult memory);
+
+  ///@notice Handle reinit request, should be implemented in Yield contract
+  /// @return reinit result
+  function _handleReinitRequest() internal virtual returns (ReinitResult memory);
+
+  ///@notice Axelar relayer calls the function when accept unstake request from Warden
```
Comment (Overseven) [inline contracts/WardenHandler.sol:170]: `/// @notice Axelar relayer calls the function when accept unstake or reinit request from Warden`

File: contracts/WardenHandler.sol [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+  struct WardenRequest {
+    uint8 actionType;
+    uint64 actionId;
+    uint128 lpAmount;
+  }
+
+  struct StakeResult {
+    uint8 status;
+    uint128 lpAmount;
+    uint128 tokenAmount;
```
Comment (Overseven) [inline contracts/WardenHandler.sol:46]: The field name is confusing, it's better to call it `reinitUnstakeTokenAmount`, imho

File: contracts/WardenHandler.sol [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+    uint64 reinitUnstakeId;
+  }
+
+  struct UnstakeResult {
+    uint8 status;
+    address tokenAddress;
+    uint128 tokenAmount;
```
Comment (Overseven) [inline contracts/WardenHandler.sol:53]: Same as in the comment above.

**Review by Overseven — CHANGES_REQUESTED** [review]

File: contracts/WardenHandler.sol (hunk 5 above)
Comment (rudewalt) [reply contracts/WardenHandler.sol:193]: left as is for now

**Review by Overseven — APPROVED** [review]

**Pull request merged**